core/store/src/trie/trie_tests.rs

```diff
@@ -202,6 +202,137 @@ mod caching_storage_tests {
         assert_eq!(get_touched_nodes_numbers(trie.clone(), state_root, &trie_items), vec![4, 4]);
 
         let storage = trie.storage.as_caching_storage().unwrap();
-        assert_eq!(storage.cache.len(), 5);
+        assert_eq!(storage.shard_cache.len(), 5);
+    }
+}
+
+#[cfg(test)]
+mod trie_cache_tests {
+    use super::*;
+    use crate::test_utils::{create_test_store, create_tries};
+    use crate::trie::trie_storage::{CachePosition, RawBytesWithCost, TrieNodeRetrievalCost};
+    use crate::trie::{TrieCache, TrieCachingStorage, TrieRefcountChange};
+    use crate::{ColState, Store, TrieChanges};
+    use assert_matches::assert_matches;
+    use near_primitives::hash::hash;
+    use near_primitives::types::TrieCacheState;
+
+    fn create_store_with_values(values: &[Vec<u8>]) -> Store {
+        let tries = create_tries();
+        let shard_uid = ShardUId::single_shard();
+        let mut trie_changes = TrieChanges::empty(Trie::empty_root());
+        trie_changes.insertions = values
+            .iter()
+            .map(|value| TrieRefcountChange {
+                trie_node_or_value_hash: hash(value),
+                trie_node_or_value: value.clone(),
+                rc: 1,
+            })
+            .collect();
+        let (store_update, _) = tries.apply_all(&trie_changes, shard_uid).unwrap();
+        store_update.commit().unwrap();
+        tries.get_store()
+    }
+
+    /// Put the item into the cache. Check that getting it from cache returns the correct value.
+    #[test]
+    fn test_put() {
+        let values = vec![vec![1u8]];
+        let store = create_store_with_values(&values);
+        let trie_cache = TrieCache::new();
+        let trie_caching_storage =
+            TrieCachingStorage::new(store, trie_cache, ShardUId::single_shard());
+        let value = &values[0];
+        let key = hash(value);
+
+        for _ in 0..2 {
+            let count_before = trie_caching_storage.get_touched_nodes_count();
+            let result = trie_caching_storage.retrieve_raw_bytes(&key);
+            let bytes: Arc<[u8]> = result.into();
+            let count_after = trie_caching_storage.get_touched_nodes_count();
+            assert_eq!(result, Some(bytes));
```

Comment:
but bytes are computed from 'result.into()'.

shouldn't you compare it with value instead?